package/gstreamer1/gst1-plugins-bad/0012-mssdemux-parse-protection-data.patch

```diff
@@ -0,0 +1,326 @@
+From 2eab71ce37ca6a77384b3cb9f9e9e7d10c1e7ccb Mon Sep 17 00:00:00 2001
+From: Charlie Turner <chturne@gmail.com>
+Date: Fri, 25 May 2018 09:43:30 +0100
+Subject: [PATCH] [mssdemux] Parse playready payload.
+
+Look at the playready specific data payload to extract default key ids
+for decrypting samples. The format of this payload seems to be
+proprietary.
+---
+ ext/smoothstreaming/gstmssdemux.c    |  20 +++-
+ ext/smoothstreaming/gstmssmanifest.c | 155 +++++++++++++++++++++++++--
+ ext/smoothstreaming/gstmssmanifest.h |   1 +
+ 3 files changed, 165 insertions(+), 11 deletions(-)
+
+diff --git a/ext/smoothstreaming/gstmssdemux.c b/ext/smoothstreaming/gstmssdemux.c
+index 26147fd8b..9298d13bb 100644
+--- a/ext/smoothstreaming/gstmssdemux.c
++++ b/ext/smoothstreaming/gstmssdemux.c
+@@ -465,7 +465,7 @@ gst_mss_demux_setup_streams (GstAdaptiveDemux * demux)
+         gst_adaptive_demux_stream_new (GST_ADAPTIVE_DEMUX_CAST (mssdemux),
+         srcpad);
+     stream->manifest_stream = manifeststream;
+-    stream->adapter = gst_adapter_new();
++    stream->adapter = gst_adapter_new ();
+     gst_mss_stream_set_active (manifeststream, TRUE);
+     active_streams = g_slist_prepend (active_streams, stream);
+   }
+@@ -508,6 +508,19 @@ gst_mss_demux_setup_streams (GstAdaptiveDemux * demux)
+           gst_event_new_protection (protection_system_id, protection_buffer,
+           "smooth-streaming");
+ 
++      GstBuffer *key_id = gst_mss_manifest_get_key_id (mssdemux->manifest);
++      if (!gst_buffer_get_size (key_id)) {
++        GST_ERROR_OBJECT (stream,
```

Comment:
My question is if erroring here is strictly necessary and if we should warn instead and let things continue. I wonder if with this change we could be breaking some already "broken but working" playback.

Reply:
My mental model here is that any protected smooth stream has to have key id(s) presentable to qtdemux, since from what i've found no such key ids exist in the media stream fetched from the server (i would like to be wrong about this, but i haven't yet found evidence to the contrary). If no key id exists, then it is known playback will fail for example. I have put a similar check in the PIFF parsing within qtdemux, but perhaps the error should ultimately happen in the decryptor when it receives no key id.

Reply:
The decryptor raising the error is what I am thinking about.